src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt

```diff
@@ -236,7 +254,29 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
     @RequiresEdt
     fun createNew(project: Project): AgentChatSession {
       val sessionId = createNewPanel(project) { it.server.chatNew() }
-      val chatSession = AgentChatSession(project, sessionId)
+      val chatSession = getSession(sessionId.get()) ?: AgentChatSession(project, sessionId)
+      synchronized(chatSessions) { chatSessions.add(chatSession) }
+      return chatSession
+    }
+
+    fun createFromState(project: Project, state: ChatState): AgentChatSession {
+      val sessionId = createNewPanel(project) { it.server.chatNew() }
+      val chatSession =
+          getSession(sessionId.get()) ?: AgentChatSession(project, sessionId, state.internalId!!)
+      for (message in state.messages) {
+        val parsed =
+            when (val speaker = message.speaker) {
+              MessageState.SpeakerState.HUMAN -> Speaker.HUMAN
+              MessageState.SpeakerState.ASSISTANT -> Speaker.ASSISTANT
+              else -> error("unrecognized speaker $speaker")
+            }
+        val chatMessage = ChatMessage(parsed, message.text, id = UUID.randomUUID())
+        chatSession.messages.add(chatMessage)
```

Comment:
You can replace those two lines with one, and simpler one:
`addMessage(chatMessage)`

Reply:
That does not make sense to me, and possible hides some other issue.
I think we need to understand what/why is happening.

Reply:
Ok, that code is fine.
The reason it is not working with just `addMessage` is that you call `HistoryService.getInstance().update(internalId, messages)` at the end of it.
You cannot iterate through history and at the same time add a new one.